### DISCUSSION TITLE
Using `implements` or `with`

### DISCUSSION THREAD

**Discussion by edwardaux** [simolus3/drift#1434] [Q&A]
Hi all, firstly, thanks for what looks to be a great project. I'm just trying to wrap my head around how to use the library properly... in particular, how to deal with subclassing / mixins.

Imagine I've got a superclass interface that looks something like:
```
abstract class Pet {
  String get name;
  int get age;
}
```
and I wanted a few of my classes to `implement` this class.
```
class Dogs extends Table {
  IntColumn get id => integer().autoIncrement()();
  TextColumn get name => text()();
  IntColumn get age => integer()();
  TextColumn get breed => text()();
}
```
How can I make the generated `Dog` class implement the `Pet` interface?  I saw #114 which looks like it was closed in favour of #1134 , however, if I understand #1134 correctly, I could create my own `Dog` class but I would also be responsible for all the boilerplate to keep it in sync with the table. Is that right?

Another approach I was considering was changing my table definition to look like:
```
@DataClassName('DogDatabase')
class Dogs extends Table {
  ...
}
```
and then declaring a new class:
```
class Dog extends DogDatabase implements Pet {
  ...
}
```
There's still challenges here, though, because functions like `copyWith()` in the generated class won't return the a `Dog`... they're going to return `DogDatabase`.

Have I missed a technique that will help here? I guess what I'm looking for is something like a `DataClassImplements` annotation that could be used to generate the appropriate code:
```
@DataClassImplements(Pet, SomeOtherInterface)
class Dogs extends Table {
}
```

**Comment by simolus3** [comment]
> I could create my own `Dog` class but I would also be responsible for all the boilerplate to keep it in sync with the table. Is that right?

Yes, although in practice it's not that much code. Moor takes care of the direction from the database to your class, you just have to add the fields as parameters to the constructor. For the other direction (e.g. for inserts or deletes), you would want to use [generated companion classes](https://moor.simonbinder.eu/docs/getting-started/writing_queries/#updates-and-deletes) instead of using the data class directly. You could also refer to other generators like [freezed](https://pub.dev/packages/freezed) or [built_value](https://pub.dev/packages/built_value) to generate those methods for custom data classes. 

Moor does not have a mechanism to implement additional interfaces in generated classes. IMO, moor-generated classes should represent a row in the database only. There's no built-in concept of inheritance in SQL, so support for those things in generated classes is limited as well. Essentially, there are two options:

- Write a manual wrapper from moor to model classes in your application
- Instruct moor to use your existing model classes with `@UseDataClass`. You still have full control over your classes with this approach, so you can add a `copyWith` method if you want to or just don't.

**Reply by edwardaux** [reply]
Thanks for the quick response. This was a key insight for me:
> moor-generated classes should represent a row in the database only

💡I think my personal challenge is that I keep trying to use Moor as an ORM, but I need to change that perspective. Moor is very focused on modelling the database and even as I type this, I can tell this even from the way the input is modelled... most other libraries (eg. Floor, ObjectBox, etc) start modelling with the entity, but Moor's input declaration is the database table  

Even though I'm personally looking for a few more ORM-type features, I also think one of the best features of Moor is that you have a very clear focus on what it should/shouldn't do. 

Thanks again for your feedback and for the library in general.
